src/cargo/ops/registry/publish.rs

```diff
@@ -115,82 +123,165 @@ pub fn publish(ws: &Workspace<'_>, opts: &PublishOpts<'_>) -> CargoResult<()> {
+    //    available (the set    we poll for is `to_confirm`).
+    //
+    // Note that `to_confirm` might contains packages from previous rounds: we
+    // might upload a bunch of packages in step 1, observe in step 2 that one
+    // has become available, and then go back to step 1 and upload a bunch more.
+    // Then `to_confirm` will contain the new packages we just uploaded, plus
+    // most of the packages from the previous round.
+    let mut order = PublishOrder::new(&pkg_dep_graph.graph);
+    let mut ready = order.ready();
+    ready.sort();
+    let mut to_confirm = BTreeSet::new();
+    let mut remaining: BTreeSet<_> = just_pkgs.iter().map(|pkg| pkg.package_id()).collect();
+
+    while !remaining.is_empty() {
+        for pkg_id in &ready {
+            let (pkg, (_features, tarball)) = &pkg_dep_graph.packages[pkg_id];
+            opts.gctx.shell().status("Uploading", pkg.package_id())?;
+
+            if !opts.dry_run {
+                let ver = pkg.version().to_string();
+
+                tarball.file().seek(SeekFrom::Start(0))?;
+                let hash = cargo_util::Sha256::new()
+                    .update_file(tarball.file())?
+                    .finish_hex();
+                let operation = Operation::Publish {
+                    name: pkg.name().as_str(),
+                    vers: &ver,
+                    cksum: &hash,
+                };
+                registry.set_token(Some(auth::auth_token(
+                    &opts.gctx,
+                    &source_ids.original,
+                    None,
+                    operation,
+                    vec![],
+                    false,
+                )?));
+            }
 
-    opts.gctx
-        .shell()
-        .status("Uploading", pkg.package_id().to_string())?;
-    transmit(
-        opts.gctx,
-        ws,
-        pkg,
-        tarball.file(),
-        &mut registry,
-        source_ids.original,
-        opts.dry_run,
-    )?;
-    if !opts.dry_run {
-        const DEFAULT_TIMEOUT: u64 = 60;
-        let timeout = if opts.gctx.cli_unstable().publish_timeout {
-            let timeout: Option<u64> = opts.gctx.get("publish.timeout")?;
-            timeout.unwrap_or(DEFAULT_TIMEOUT)
+            transmit(
+                opts.gctx,
+                ws,
+                pkg,
+                tarball.file(),
+                &mut registry,
+                source_ids.original,
+                opts.dry_run,
+            )?;
+
+            if !opts.dry_run {
+                // Short does not include the registry name.
+                let short_pkg_description = format!("{} v{}", pkg.name(), pkg.version());
+                let source_description = source_ids.original.to_string();
+                ws.gctx().shell().status(
+                    "Uploaded",
+                    format!("{short_pkg_description} to {source_description}"),
+                )?;
+            }
+        }
+
+        to_confirm.extend(ready.iter().copied());
```

Comment:
When reading this, it implies to me that `ready` will continue to be used.

Some ideas
- `to_confirm.extend(ready.drain())` (depending on `ready`s type)
- `to_confirm.extend(std::mem::take(&mut ready)` 
- Inserting into `to_confirm` on the line right after `transmit`

... wait, does `ready` need to cross iterations of the loop?  We could loop over `remaining` and then immediately create `ready` before looping over it.  I feel like that would be a big help for understanding the code.

very minor nit: If we don't go with the last option, we *might* want to move `to_confirm` up a line at least.  Blank lines in code convey logical grouping.  Seeing a line like this suggests to me that we are initializing `to_confirm` for the following line but its for that and for future iterations of the loop.  Attaching it to the transmit operation (directly after it or at least after the block without a newline) conveys that what we transmit must be confirmed at some point.

Reply:
I went with the last option.

I also pushed some of the state into `PublishOrder` so that the loop is hopefully clearer. It would be also be possible to remove some of that state altogether, but (AFAICS) only at the cost of making things quadratic in the number of packages. (For example, we could remove ready, and just recalculate it every time as the set of remaining packages with weight zero.) Maybe being quadratic in the number of packages isn't important? But on the other hand I think the state is reasonably well-encapsulated now...